# Jobs, Jobs, jobs! where is easier to find it?
<br>


I will look between the relation of jobs available in each country against the population of that country to find the highest country with jobs available for person and increase the chances of finding that desired and dreamed perfect job!!!.
<br>

I will use the library selenium along with Pandas to get the data from indeed looking in each indeed from each country respectively and clean the data lets dive in!!



we will start by import the libraries and inspect the web page that we are going to get the data from
<bre>

In [5]:
from selenium import webdriver
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
import time

In [6]:
# # this option is to keep the browser open after the program runs
# options = webdriver.ChromeOptions()
# options.add_experimental_option("detach", True)

# # Chromeedrivermanager will confg the path
# driver = webdriver.Chrome(options=options, executable_path=ChromeDriverManager().install())

In [31]:
from selenium.webdriver.chrome.service import Service
driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()))

In [32]:
url='https://www.indeed.com/worldwide?from=mobRdr&utm_source=%2Fm%2F&utm_medium=redir&utm_campaign=dt'
driver.get(url)
driver.find_element(By.ID,'onetrust-reject-all-handler').click()

we need the list of links that will be used for the driver to scrape the data we need,in this case that list is stored in 'countryLinks' and 'listatrabajos' will be the list of diccionaries where we record the name of the country and the jobs available in each of them


In [33]:
# iterate to get the population and jobs number
countryLinks = driver.find_elements(By.XPATH,'//a[@class="worldwide__link"]')
listatrabajos=[]

try and except are needed to block the pop-ups and add other routes because each website varies slightly, this is the best way to avoid a bug in the code.

In [36]:
for link in countryLinks:
        driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()))
        cname=link.text
        driver.get(str(link.get_attribute('href')))
        time.sleep(2)
        try:
            driver.find_element(By.ID,'onetrust-reject-all-handler').click()
        except:
                pass
        time.sleep(2)
        try:
                driver.find_element(By.XPATH,'//button[@class="icl-CloseButton icl-Card-close"]').click()      #cierra el boton de pop u de google
        except:
                pass
        donde=driver.find_element(By.XPATH,'//div[@class="yosegi-EmbeddedLabelInput yosegi-InlineWhatWhere-where"]')
        donde.click()
        driver.find_element(By.XPATH,'//input[@id="text-input-where"]').send_keys(cname)
        driver.find_element(By.XPATH,'//input[@id="text-input-where"]').send_keys(Keys.ENTER)
        time.sleep(2)
        try:
                try:
                        jobscount=driver.find_element(By.XPATH,'//div[@id="searchCount"]')
                except:
                        jobscount=driver.find_element(By.XPATH,'//div[@class="jobsearch-JobCountAndSortPane-jobCount"]')
        except:
                pass
        try:
                jobscount=jobscount.text
        except:
                pass
        try:
                jobscount=jobscount.split(' ')[-2]
        except:
                pass
        try:
                print(jobscount+' '+cname)
                diccionario={'jobs':jobscount,'pais':cname}
                
                listatrabajos.append(diccionario)
        except:
                pass
# The output below shows the result of the code

24,420 Argentina
174,941 Australia
65,380 Austria
1,248 Bahrain
239,588 Belgium
335.733 Brazil
502,871 Canada
84.343 Chile
131,804个工作，第1页 China
33.083 Colombia
7,444 Costa Rica
306 Czech Republic
40.788 Denmark
40.788 Ecuador
40.788 Egypt
702 Finland
702 France
702 Germany
θέσεις Greece
θέσεις Hong Kong
1 Hungary
188,572 India
188,572 Indonesia
58,180 Ireland
68,975 Israel
200.286 Italy
1 Japan
2,384 Kuwait
15,311 Luxembourg
130,569 Malaysia
244,730 Mexico
121 Morocco
437.856 Netherlands
437.856 New Zealand
6,784 Nigeria
299 Norway
1,524 Oman
21,604 Pakistan
3,087 Panama
24,649 Peru
155,045 Philippines
99,841 Poland
49.605 Portugal
11,498 Qatar
de Romania
15,993 Saudi Arabia
157,992 Singapore
32,091 South Africa
결과 South Korea
95.400 Spain
423 Sweden
75,665 Switzerland
52,498 Taiwan
124,110 Thailand
30.244 Turkey
118 Ukraine
28,655 United Arab Emirates
737,410 United Kingdom
737,410 United States
737,410 Uruguay
3.186 Venezuela
77,624 Vietnam


Now we start with the data cleaning as we can see there are some missing values like Hungary,Greece,Hong Kong etc... that we will introduce manualy,moreover,the population of each country needs to be added aswell.


In [2]:
import pandas as pd

In [215]:
data=pd.DataFrame(data=listatrabajos)
data=data.rename(columns={'pais':'Country'})

In [216]:
data.head()

jobs    Country
0   24,420  Argentina
1  174,941  Australia
2   65,380    Austria
3    1,248    Bahrain
4  239,588    Belgium

In [217]:
# Fixig those values that were odd manually
data.loc[18]['jobs']=14850
data.loc[19]['jobs']=77451
data.loc[8]['jobs']=131804
data.loc[16]['jobs']=210042
data.loc[20]['jobs']=27000
data.loc[15]['jobs']=13627
data.loc[17]['jobs']=1402408
data.loc[26]['jobs']=6542747
data.loc[35]['jobs']=49547
data.loc[44]['jobs']=32190
data.loc[48]['jobs']=131321
data.loc[50]['jobs']=72711
data.loc[58]['jobs']=7352938
data.loc[11]['jobs']=50473
data.loc[55]['jobs']=20126
data.loc[59]['jobs']=3205
data.loc[33]['jobs']=33000

In [218]:
# this i the dataframe resulting from the web scrape and the data wrangling.
data

jobs         Country
0    24,420       Argentina
1   174,941       Australia
2    65,380         Austria
3     1,248         Bahrain
4   239,588         Belgium
..      ...             ...
57  737,410  United Kingdom
58  7352938   United States
59     3205         Uruguay
60    3.186       Venezuela
61   77,624         Vietnam

[62 rows x 2 columns]

In [231]:
data.to_csv('/Users/ruimandiaz/Documents/jobsworldwide.csv')

In this case I have used [this data](https://worldpopulationreview.com/) where I found the population of each country, we just need to merge the two dataframe 

In [13]:
Population=pd.read_csv('csvData.csv')

Unemployment table below gives us the employment rate and hence the unemployment rate which is used calculated how many population is without a job in each country

In [75]:
unemployment=pd.read_excel('unemployment_rate.xlsx')

In [76]:
unemployment.head()

Country  Employment_rate  Unemployment_rate  rate_percentage
0       Mexico            96.57               3.43           0.0343
1     SriLanka            95.40               4.60           0.0460
2  Philippines            94.00               6.00           0.0600
3      Jamaica            94.00               6.00           0.0600
4    Venezuela            93.65               6.35           0.0635

In [233]:
Population=Population[['name','pop2022']]
Population=Population.rename(columns={'name':'Country','pop2022':'Population'})

In [234]:
wdata=pd.merge(data,Population,on='Country')

In [235]:
wdata['jobs']=[str(x) if type(x)==int else x for x in wdata['jobs'] ]

In [236]:
wdata['jobs']=[x.replace(',','').replace('.','') for x in wdata['jobs']]

In [237]:
wdata['jobs']=[int(x) for x in wdata['jobs']]

In [77]:
wdata=pd.merge(data,unemployment,on='Country')

In [79]:
wdata['person/job']=round((wdata['Population'])*1000*wdata['rate_percentage']/wdata['jobs'])

In [89]:
wdata

jobs         Country   Population  Employment_rate  Unemployment_rate  \
0     24420       Argentina    45510.318            43.30              56.70   
1    174941       Australia    26177.413            64.28              35.72   
2     65380         Austria     8939.617            74.10              25.90   
3    239588         Belgium    11655.930            66.00              34.00   
4    335733          Brazil   215313.498            57.00              43.00   
5    502871          Canada    38454.327            61.30              38.70   
6     84343           Chile    19603.733            55.10              44.90   
7     33083        Colombia    51874.024            56.53              43.47   
8      7444      Costa Rica     5180.829            52.62              47.38   
9     50473  Czech Republic    10493.986            75.20              24.80   
10    40788         Denmark     5882.261            67.30              32.70   
11    40788         Ecuador    18001.000            42.00              58.00   
12    13627         Finland     5540.745            75.70              24.30   
13   210042          France    64626.628            68.00              32.00   
14  1402408         Germany    83369.843            77.30              22.70   
15    14850          Greece    10384.971            88.46              11.54   
16    27000         Hungary     9967.308            64.11              35.89   
17   188572           India  1417173.173            42.30              57.70   
18    58180         Ireland     5023.109            73.50              26.50   
19    68975          Israel     9038.309            60.60              39.40   
20   200286           Italy    59037.474            59.90              40.10   
21  6542747           Japan   123951.692            61.30              38.70   
22    15311      Luxembourg      647.599            70.30              29.70   
23   244730          Mexico   127504.125            96.57               3.43   
24      121         Morocco    37457.971            40.20              59.80   
25   437856     Netherlands    17564.014            81.80              18.20   
26    33000     New Zealand     5185.288            68.50              31.50   
27     6784         Nigeria   218541.212            66.70              33.30   
28    49547          Norway     5434.319            70.70              29.30   
29    24649            Peru    34049.588            92.70               7.30   
30   155045     Philippines   115559.009            94.00               6.00   
31    99841          Poland    39857.145            71.40              28.60   
32    49605        Portugal    10270.865            56.40              43.60   
33    32190         Romania    19659.267            67.80              32.20   
34    32091    South Africa    59893.885            38.70              61.30   
35   131321     South Korea    51815.810            61.40              38.60   
36    95400           Spain    47558.630            51.38              48.62   
37    72711          Sweden    10549.347            68.50              31.50   
38    75665     Switzerland     8740.472            79.30              20.70   
39    30244          Turkey    85341.241            46.50              53.50   
40    20126         Ukraine    39701.739            64.10              35.90   
41   737410  United Kingdom    67508.936            75.40              24.60   
42  7352938   United States   338289.857            60.10              39.90   
43     3205         Uruguay     3422.794            56.80              43.20   
44     3186       Venezuela    28301.696            93.65               6.35   

    rate_percentage  person/job  
0            0.5670      1057.0  
1            0.3572        53.0  
2            0.2590        35.0  
3            0.3400        17.0  
4            0.4300       276.0  
5            0.3870        30.0  
6            0.4490       104.0  
7            0.4347       682.0  
8            0.4738 

### BOTTOM 15 COUNTRIES with highest person per job are:

In [90]:
wdata.sort_values('person/job',ascending=False,ignore_index=True).head(15)

jobs       Country   Population  Employment_rate  Unemployment_rate  \
0      121       Morocco    37457.971            40.20              59.80   
1     6784       Nigeria   218541.212            66.70              33.30   
2   188572         India  1417173.173            42.30              57.70   
3    30244        Turkey    85341.241            46.50              53.50   
4    32091  South Africa    59893.885            38.70              61.30   
5    24420     Argentina    45510.318            43.30              56.70   
6    20126       Ukraine    39701.739            64.10              35.90   
7    33083      Colombia    51874.024            56.53              43.47   
8     3186     Venezuela    28301.696            93.65               6.35   
9     3205       Uruguay     3422.794            56.80              43.20   
10    7444    Costa Rica     5180.829            52.62              47.38   
11  335733        Brazil   215313.498            57.00              43.00   
12   40788       Ecuador    18001.000            42.00              58.00   
13   95400         Spain    47558.630            51.38              48.62   
14   32190       Romania    19659.267            67.80              32.20   

    rate_percentage  person/job  
0            0.5980    185123.0  
1            0.3330     10727.0  
2            0.5770      4336.0  
3            0.5350      1510.0  
4            0.6130      1144.0  
5            0.5670      1057.0  
6            0.3590       708.0  
7            0.4347       682.0  
8            0.0635       564.0  
9            0.4320       461.0  
10           0.4738       330.0  
11           0.4300       276.0  
12           0.5800       256.0  
13           0.4862       242.0  
14           0.3220       197.0

In this case we could observ that in the case of Morocco it has 185,123 person for each job available and so on.

### TOP 15 COUNTRIES with the lowest persons per job are:

In [91]:
wdata.sort_values('person/job',ascending=True,ignore_index=True).head(15)

jobs         Country  Population  Employment_rate  Unemployment_rate  \
0    437856     Netherlands   17564.014            81.80              18.20   
1   6542747           Japan  123951.692            61.30              38.70   
2     15311      Luxembourg     647.599            70.30              29.70   
3   1402408         Germany   83369.843            77.30              22.70   
4    239588         Belgium   11655.930            66.00              34.00   
5    244730          Mexico  127504.125            96.57               3.43   
6   7352938   United States  338289.857            60.10              39.90   
7     58180         Ireland    5023.109            73.50              26.50   
8    737410  United Kingdom   67508.936            75.40              24.60   
9     75665     Switzerland    8740.472            79.30              20.70   
10   502871          Canada   38454.327            61.30              38.70   
11    49547          Norway    5434.319            70.70              29.30   
12    65380         Austria    8939.617            74.10              25.90   
13   155045     Philippines  115559.009            94.00               6.00   
14    72711          Sweden   10549.347            68.50              31.50   

    rate_percentage  person/job  
0            0.1820         7.0  
1            0.3870         7.0  
2            0.2970        13.0  
3            0.2270        13.0  
4            0.3400        17.0  
5            0.0343        18.0  
6            0.3990        18.0  
7            0.2650        23.0  
8            0.2460        23.0  
9            0.2070        24.0  
10           0.3870        30.0  
11           0.2930        32.0  
12           0.2590        35.0  
13           0.0600        45.0  
14           0.3150        46.0

In this case this are the better countries to find a job as the count of person for job is less, in the case of netherlands and japan we have 7 persons for each job available in indeed.

In [92]:
wdata.to_csv('/Users/ruimandiaz/Documents/Worldwidejobsperperson.csv')

Disclaimer: This analisys only have in consideration the webpage indeed and nothing more, the count of job posting of every country is more elaborated and complex than this.There is only 61 countries in this analisys.